include/StepRecorder.h

```diff
@@ -0,0 +1,143 @@
+/*
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef STEP_RECORDER_H
+#define STEP_RECORDER_H
+
+#include <QTime>
+#include <QTimer>
+#include <QObject>
+#include <QKeyEvent>
+
+#include "Note.h"
+#include "lmms_basics.h"
+#include "Pattern.h"
+
+class PianoRoll;
+class StepRecorderWidget;
+
+class StepRecorder : public QObject
+{
+	Q_OBJECT
+
+	public:
+	StepRecorder(PianoRoll& pianoRoll, StepRecorderWidget& stepRecorderWidget);
+
+	void initialize();
+	void start(const MidiTime& currentPosition,const MidiTime& stepLength);
+	void stop();     
```

Comment:
Some more hidden trailing whitespaces here. You may want to find some more occurrences before merging this.
Anyway, you don't need to worry about this for now. :smiley:

Reply:
fixed